py/server/deephaven/table.py

```diff
@@ -103,66 +101,47 @@ def _encode_signature(fn: Callable) -> str:
     return "".join(np_type_codes)
 
 
-class DhVectorize:
+def _dh_vectorize(fn):
     """A decorator to vectorize a Python function used in Deephaven query formulas and invoked on a row basis.
 
-    The result callable will accept arrays of the declared parameter data types (vs. scalar) and iterate through them
-    together and call the wrapped function on each iteration and pack the result of each call into an array and return
-    the result array upon exhausting the array parameters.
+    The result wrapper accepts arrays of the declared parameter data types (vs. scalar), iterates through them
+    together, and calls the wrapped function on each iteration. The result of each call is packed into an array which
+    is then returned when the input arrays are exhausted. In addition, the result wrapper expects an extra
+    'chunk-size' parameter that the caller must provide to determine the actual size of the input arrays. Since this
+    potentially changes the signature of the wrapped function, it is not recommended for the user to use this
+    decorator with the @ syntax.
     """
+    signature = _encode_signature(fn)
+    return_type = signature[-1]
+
+    def wrapper(*args):
+        if len(args) != len(signature) - len("->?") + 1:
+            raise ValueError(
+                f"The number of arguments doesn't match the function signature. {len(args) - 1}, {signature}")
+        if args[0] <= 0:
+            raise ValueError(f"The chunk size argument must be a positive integer. {args[0]}")
+
+        chunk_size = args[0]
+        chunk_result = np.empty(chunk_size, return_type)
```

Comment:
I would love to be able to re-use return arrays, and just wrap them as np arrays.